src/lib/prov/tpm2/tpm2_crypto_backend/tpm2_crypto_backend.cpp

```diff
@@ -0,0 +1,859 @@
+
+      const auto pubkey = Botan::TPM2::rsa_pubkey_from_tss2_public(pub_tpm_key);
+      const auto keybits = pubkey.key_length();
+      const auto output_size = keybits / 8;
+      if(eme->maximum_input_size(keybits) < in_size) {
+         return TSS2_ESYS_RC_BAD_VALUE;
+      }
+
+      if(output_size > max_out_size) {
+         return TSS2_ESYS_RC_INSUFFICIENT_BUFFER;
+      }
+
+      auto ccs = get(userdata);
+      if(!ccs) {
+         return TSS2_ESYS_RC_BAD_REFERENCE;
+      }
+
+      Botan::RandomNumberGenerator& rng = *ccs->get().rng;
+
+      const auto max_raw_bits = keybits - 1;
+      const auto max_raw_bytes = (max_raw_bits + 7) / 8;
+      const auto padded_bytes = eme->pad({out_buffer, max_raw_bytes}, {in_buffer, in_size}, max_raw_bits, rng);
+
+      // PK_Encryptor_EME does not provide a way to pass in an output buffer.
+      // TODO: provide an `.encrypt()` overload that accepts an output buffer.
+      Botan::PK_Encryptor_EME encryptor(pubkey, rng, "Raw");
+      const auto encrypted = encryptor.encrypt({out_buffer, padded_bytes}, rng);
+
+      // We abused the `out_buffer` to hold the result of the padding. Hence, we
+      // now have to copy the encrypted data over the padded plaintext data.
+      *out_size = encrypted.size();
+      Botan::copy_mem(std::span{out_buffer, *out_size}, encrypted);
+
+      return TSS2_RC_SUCCESS;
+   });
+   #else
+   BOTAN_UNUSED(pub_tpm_key, in_size, in_buffer, max_out_size, out_buffer, out_size, label, userdata);
+   return TSS2_ESYS_RC_NOT_IMPLEMENTED;
+   #endif
+}
+
+/** Computation of an ephemeral ECC key and shared secret Z.
+ *
+ * According to the description in TPM spec part 1 C 6.1 a shared secret
+ * between application and TPM is computed (ECDH). An ephemeral ECC key and a
+ * TPM key are used for the ECDH key exchange.
+ * @param[in] key The key to be used for ECDH key exchange.
+ * @param[in] max_out_size the max size for the output of the public key of the
+ *            computed ephemeral key.
+ * @param[out] Z The computed shared secret.
+ * @param[out] Q The public part of the ephemeral key in TPM format.
+ * @param[out] out_buffer The public part of the ephemeral key will be marshaled
+ *             to this buffer.
+ * @param[out] out_size The size of the marshaled output.
+ * @param[in,out] userdata information.
+ * @retval TSS2_RC_SUCCESS on success
+ * @retval USER_DEFINED user defined errors on failure.
+ */
+static TSS2_RC get_ecdh_point(TPM2B_PUBLIC* key,
+                              size_t max_out_size,
+                              TPM2B_ECC_PARAMETER* Z,
+                              TPMS_ECC_POINT* Q,
+                              BYTE* out_buffer,
+                              size_t* out_size,
+                              void* userdata) {
+   #if defined(BOTAN_HAS_TPM2_ECC_ADAPTER)
+   return thunk([&] {
+      BOTAN_UNUSED(userdata);
+      BOTAN_ASSERT_NONNULL(key);
+      BOTAN_ASSERT_NOMSG(key->publicArea.type == TPM2_ALG_ECC);
```

Comment:
Not sure, but I believe this could also be `TPM2_ALG_SM2` if we deal with an SM2 key... We should figure out a way to test this.

Reply:
Actually, `type = TPM2_ALG_ECC` for all SM2/ECDSA/ECDH keys. `scheme` will be different if the TPM key is restricted to a certain scheme like SM2/ECDSA/ECDH.

Starting a session with an explicit ECDH key works since the `decrypt` flag can be set for this scheme. However, if we restrict the scheme to SM2 or ECDSA, `tpm2-tools`does not allow to set the decrypt flag since these are signature  schemes.

=> The assert is fine, however we cannot test the backend with SM2 since we currently build out sessions to require the `decrypt` key object attribute.

I removed this assert because it already made when loading the key as a `Botan::TPM2::EC_PublicKey`.